### PR TITLE
Custom Entity Tracking

### PR DESCRIPTION
**Pull request by flusedev**
Adds a config for changing the entity tracking range based on its type.

### CODE REVIEW

**Comment by HookWoods** [comment]
Why creating a new config ? You could just simply port it to yatopia existing config. And its yatopia, not your own fork so your custom config must be in yatopia folder

**Comment by toinouH** [comment]
@fluse1367 Join our Discord to talk with us: https://discord.io/YatopiaMC

**Review by MrIvanPlays — CHANGES_REQUESTED** [review]
definite no for the new config. move the values into yatopia.yml

**Comment by Titaniumtown** [comment]
yeaaaaa, why is this in a new config lol

**Comment by MrIvanPlays** [comment]
also every new file should go under the org.yatopiamc.yatopia package

**Comment by flusedev** [comment]
Entity tracking ranges can now be controlled via the API:
 - EntityType#getTrackingDistance
 - EntityType#getTrackingDistance(World)
 - EntityType#setTrackingDistance(int)
 - EntityType#setTrackingDistance(World, int)
 - EntityType#resetTrackingDistance
 - EntityType#resetTrackingDistance(World)
 - Entity#getTrackingDistance
 - Entity#setTrackingDistance(int)
 - Entity#resetTrackingDistance

The custom tracking distances in Entity have higher priority than the ones in EntityType. If nothing's set (which is the default) the regular (spigot determined) tracking distances are used.

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+--- a/src/main/java/org/bukkit/entity/Entity.java
++++ b/src/main/java/org/bukkit/entity/Entity.java
+@@ -719,5 +719,23 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
+     default de.tr7zw.changeme.nbtapi.NBTCompound getNBTC() {
+         return getNBT().getPersistentDataContainer();
+     }
++
++    /**
++     * Returns the tracking (render) distance for this specific {@link Entity},<br>
++     *      *     falls back to {@link EntityType#getTrackingDistance(World)} if not set.
```
Comment (HookWoods) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch]: Code style error

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+@@ -425,4 +427,60 @@ public enum EntityType implements Keyed {
+     String getTranslationKey() {
+         return org.bukkit.Bukkit.getUnsafe().getTranslationKey(this);
+     }
++
++    // Yatopia start
++
++    /**
++     * Returns the tracking (render) distance for a specific {@link EntityType}, {@code -1} if not set.
++     * @return the tracking (render) distance for a specific {@link EntityType}, {@code -1} if not set.
```
Comment (HookWoods) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch:57]: Why having it two times

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
++     */
++    public int getTrackingDistance() {
++        return trackingDistance;
++    }
++
++    /**
++     * Returns the tracking (render) distance for a specific {@link EntityType} in the {@link World} {@code world},<br>
++     *     falls back to {@link #getTrackingDistance()} if not set.
```
Comment (HookWoods) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch:65]: here too

File: patches/server/0066-Per-Entity-Tracking-Ranges.patch [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+@@ -1160,4 +1161,23 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+         return getHandle().isTicking();
+     }
+     // Paper end
++
++    // Yatopia start
++
++    @Override
++    public int getTrackingDistance() {
++        return trackingDistance < 0 ? getType().getTrackingDistance(getWorld()) : trackingDistance;
```
Comment (HookWoods) [inline patches/server/0066-Per-Entity-Tracking-Ranges.patch:28]: What happens if the tracking distance is equal to 0 for the system?

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch (hunk 1 above)
Comment (MrIvanPlays) [reply patches/api/0009-Per-Entity-Tracking-Ranges.patch]: fix it

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch (hunk 2 above)
Comment (MrIvanPlays) [reply patches/api/0009-Per-Entity-Tracking-Ranges.patch:57]: Returns the tracking distance for a specific `{@link EntityType}, <code>-1</code>` if not set.

`@return the tracking distance`

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch (hunk 3 above)
Comment (MrIvanPlays) [reply patches/api/0009-Per-Entity-Tracking-Ranges.patch:65]: Returns the tracking distance for a specific `{@link EntityType} in the specified {@link World}`
`<p>`
Falls back to `{@link #getTrackingDistance()}` if not set.

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
++     * @return Returns the tracking (render) distance for a specific {@link EntityType} in the {@link World} {@code world}.
```
Comment (MrIvanPlays) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch:66]: `@return the tracking distance in the specified world`

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
++    /**
++     * Sets the tracking (render) distance for a specific {@link EntityType}.
++     * @param distance the tracking distance to set.
++     */
++    public void setTrackingDistance(int distance) {
++        this.trackingDistance = distance;
++    }
++
++    /**
++     * Sets the tracking (render) distance for a specific {@link EntityType} in the {@link World} {@code world}.
```
Comment (MrIvanPlays) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch:81]: remove (render) and `{@code world}`

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
++    public void setTrackingDistance(@NotNull World world, int distance) {
++        if (distance < 0) {
++            resetTrackingDistance(world);
++        } else {
++            trackingDistances.put(world, distance);
++        }
++    }
++
++    /**
++     * Sets the tracking (render) distance for a specific {@link EntityType} back to default.
```
Comment (MrIvanPlays) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch:94]: remove (render)

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
++     */
++    public void resetTrackingDistance() {
++        this.trackingDistance = -1;
++    }
++
++    /**
++     * Sets the tracking (render) distance for a specific {@link EntityType} in the {@link World} {@code world} back to default.
```
Comment (MrIvanPlays) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch:101]: remove (render) and `{@code world}`

File: patches/server/0066-Per-Entity-Tracking-Ranges.patch [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+index a6d849facba1526ae2a2b7f3fb9a140d0b50289c..b4df64cfc15fce322849f13138973ffa6dbc5f5a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+@@ -181,6 +181,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+     protected Entity entity;
+     private EntityDamageEvent lastDamageEvent;
+     private final CraftPersistentDataContainer persistentDataContainer = new CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
++    private int trackingDistance = -1;
```
Comment (MrIvanPlays) [inline patches/server/0066-Per-Entity-Tracking-Ranges.patch:15]: `private int trackingDistance = -1; // Yatopia`

File: patches/server/0066-Per-Entity-Tracking-Ranges.patch (hunk 4 above)
Comment (MrIvanPlays) [reply patches/server/0066-Per-Entity-Tracking-Ranges.patch:28]: perhaps a validation to make sure it is never 0

File: patches/server/0066-Per-Entity-Tracking-Ranges.patch [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+diff --git a/src/main/java/org/spigotmc/TrackingRange.java b/src/main/java/org/spigotmc/TrackingRange.java
+index 5cfe16ef8c5a7b4405efa58b8fc532e14353f88d..f3255829cef9ffc020acc5774b87c1f11c40f7c8 100644
+--- a/src/main/java/org/spigotmc/TrackingRange.java
++++ b/src/main/java/org/spigotmc/TrackingRange.java
+@@ -23,6 +23,10 @@ public class TrackingRange
+      */
+     public static int getEntityTrackingRange(Entity entity, int defaultRange)
+     {
++        int trackingDistance = entity.getBukkitEntity().getTrackingDistance();
++        if (trackingDistance >= 0) {
```
Comment (MrIvanPlays) [inline patches/server/0066-Per-Entity-Tracking-Ranges.patch:52]: trackingDistance > 0

File: patches/server/0066-Per-Entity-Tracking-Ranges.patch [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
++    // Yatopia start
++
++    @Override
++    public int getTrackingDistance() {
++        return trackingDistance < 0 ? getType().getTrackingDistance(getWorld()) : trackingDistance;
++    }
++
++    @Override
++    public void setTrackingDistance(int distance) {
++        this.trackingDistance = Math.max(-1, distance);
```
Comment (MrIvanPlays) [inline patches/server/0066-Per-Entity-Tracking-Ranges.patch:33]: see comment on line 28 - add a validation to make sure it is never 0
`Preconditions.checkArgument(distance > 0, "The set tracking distance should be bigger than 0");`

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
++     */
++    public int getTrackingDistance(@NotNull World world) {
++        return trackingDistances.getOrDefault(world, getTrackingDistance());
++    }
++
++    /**
++     * Sets the tracking (render) distance for a specific {@link EntityType}.
++     * @param distance the tracking distance to set.
++     */
++    public void setTrackingDistance(int distance) {
```
Comment (MrIvanPlays) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch:76]: Add a validation to make sure distance is never 0

**Review by MrIvanPlays — COMMENTED** [review]
I like that implementation. There are some minor things here and there
What happened to the config btw? You don't want both an API and a config?

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
+index 692b75eb78405874077c850bfc72e247ccc80860..3def6f19cdd3acf51c1ca96385e7c78c3a392e2f 100644
+--- a/src/main/java/org/bukkit/entity/EntityType.java
++++ b/src/main/java/org/bukkit/entity/EntityType.java
+@@ -287,6 +287,8 @@ public enum EntityType implements Keyed {
+     private final short typeId;
+     private final boolean independent, living;
+     private final NamespacedKey key;
++    private int trackingDistance = -1; // Yatopia
++    private final Map<World, Integer> trackingDistances = new HashMap<>(); // Yatopia
```
Comment (HookWoods) [inline patches/api/0009-Per-Entity-Tracking-Ranges.patch:44]: Maybe directly use Fastutil? We have it so why not use a very small performance enhancement :)

**Review by HookWoods — CHANGES_REQUESTED** [review]

File: patches/api/0009-Per-Entity-Tracking-Ranges.patch (hunk 13 above)
Comment (duplexsystem) [reply patches/api/0009-Per-Entity-Tracking-Ranges.patch:44]: Not always. Fastutil and similar alternatives to Java's built in HashSet, HashMap, and other collections are primarily designed with reducing memory footprint in mind. Although in some instances they offer performance gains varying in size a majority of the time, especially with HashMap alternatives they greatly reduce performance in most cases. I suggest you take a look at this paper: https://research.spec.org/icpe_proceedings/2017/proceedings/p389.pdf

**Comment by toinouH** [comment]
What are you doing @Titaniumtown ?

**Comment by Titaniumtown** [comment]
I didn't close the PR.... prob something it #490's description.

**Comment by flusedev** [comment]
I abandoned yatopia as a server software, i won't work on this commit anymore anywys.

**Pull request closed without merging**